### PR TITLE
Changes to support code patterns in the existing codebase

### PR DESCRIPTION
**Pull request by toddlang**
-Added tests that emulate structure of code that failed previously

### CODE REVIEW

File: src/D2L.CodeStyle.Analyzers/Immutability/MutabilityInspector.cs [hunk 1]
```diff
@@ -126,15 +126,47 @@ HashSet<ITypeSymbol> typeStack
 			// sufficiently immutable, and we also need to ensure the
 			// immutability of the type parameters is also appropriate.
 			ImmutabilityScope scope = type.GetImmutabilityScope();
-			if( !type.IsGenericType()
-				&& !flags.HasFlag( MutabilityInspectionFlags.IgnoreImmutabilityAttribute )
+			if( !flags.HasFlag( MutabilityInspectionFlags.IgnoreImmutabilityAttribute )
 				&& scope != ImmutabilityScope.None
 			) {
+				if( type.IsGenericType() ) {
```
Comment (toddlang) [inline src/D2L.CodeStyle.Analyzers/Immutability/MutabilityInspector.cs]: We get in here if it's marked immutable, so now we need to perform the additional check on the type parameters that we weren't before.  Specifically, check if the parameter is supposed to be immutable, then check if the type evaluates to immutable.

File: src/D2L.CodeStyle.Analyzers/Immutability/MutabilityInspector.cs [hunk 2]
```diff
@@ -126,15 +126,47 @@ HashSet<ITypeSymbol> typeStack
...
+				}
+
+				// If we didn't bail before here then all the type arguments
+				// are valid, so it's not mutable-ish.
 				ImmutableHashSet<string> immutableExceptions = type.GetAllImmutableExceptions();
 				return MutabilityInspectionResult.NotMutable( immutableExceptions );
 			}
 
 			return DoInspectType(
+				type,
```
Comment (toddlang) [inline src/D2L.CodeStyle.Analyzers/Immutability/MutabilityInspector.cs]: This doesn't hurt.  We could also easily pass default for the "container" type as well.

File: src/D2L.CodeStyle.Analyzers/Immutability/MutabilityInspector.cs [hunk 3]
```diff
@@ -406,61 +491,12 @@ HashSet<ITypeSymbol> typeStack
 
 			return DoInspectType(
 				field.Type,
+				hostSymbol,
 				MutabilityInspectionFlags.Default,
 				typeStack
 			).WithPrefixedMember( field.Name );
 		}
 
-		/// <summary>
```
Comment (toddlang) [inline src/D2L.CodeStyle.Analyzers/Immutability/MutabilityInspector.cs]: Just moved to the bottom of the file.

File: src/D2L.CodeStyle.Analyzers/Immutability/MutabilityInspector.cs [hunk 4]
```diff
@@ -126,15 +126,47 @@ HashSet<ITypeSymbol> typeStack
...
-			if( !type.IsGenericType()
-				&& !flags.HasFlag( MutabilityInspectionFlags.IgnoreImmutabilityAttribute )
+			if( !flags.HasFlag( MutabilityInspectionFlags.IgnoreImmutabilityAttribute )
 				&& scope != ImmutabilityScope.None
 			) {
+				if( type.IsGenericType() ) {
+					// This is some kind of immutable generic class, now we 
+					// need to check that the types specified are appropriate
+					var genericSymbol = type as INamedTypeSymbol;
+					foreach( ITypeSymbol typeArgument in genericSymbol.TypeArguments ) {
```
Comment (JeffAshton) [inline src/D2L.CodeStyle.Analyzers/Immutability/MutabilityInspector.cs]: Null reference exception

File: src/D2L.CodeStyle.Analyzers/Immutability/MutabilityInspector.cs (hunk 2 above)
Comment (JeffAshton) [reply src/D2L.CodeStyle.Analyzers/Immutability/MutabilityInspector.cs]: name these parameters

File: src/D2L.CodeStyle.Analyzers/Immutability/MutabilityInspector.cs [hunk 5]
```diff
@@ -161,6 +193,7 @@ HashSet<ITypeSymbol> typeStack
 
 				MutabilityInspectionResult result = DoInspectType(
 					arg,
+					type,
```
Comment (JeffAshton) [inline src/D2L.CodeStyle.Analyzers/Immutability/MutabilityInspector.cs]: Name these

File: src/D2L.CodeStyle.Analyzers/Immutability/MutabilityInspector.cs (hunk 3 above)
Comment (JeffAshton) [reply src/D2L.CodeStyle.Analyzers/Immutability/MutabilityInspector.cs]: Looks like you changed the method too.  Did you mean to?

image

Comment (JeffAshton) [reply src/D2L.CodeStyle.Analyzers/Immutability/MutabilityInspector.cs]: Actually looks like a merge conflict.   Do smaller PRs.  Move methods on their own.  Not as a part of other PRs

**Pull request closed without merging**